ComponentKit/TransactionalDataSources/Common/Internal/CKTransactionalComponentDataSourceChangesetVerification.mm

```diff
@@ -146,6 +165,85 @@ CKBadChangesetOperationType CKIsValidChangesetForState(CKTransactionalComponentD
+    [modifiedSections addObject:[items mutableCopy]];
+  }];
+  // Update items
+  [[changeset updatedItems] enumerateKeysAndObjectsUsingBlock:^(NSIndexPath *indexPath, id model, BOOL *stop) {
+    [modifiedSections[indexPath.section] replaceObjectAtIndex:indexPath.item withObject:model];
+  }];
+  __block std::unordered_map<NSUInteger, std::map<NSUInteger, id>> insertedItemsBySection;
+  __block std::unordered_map<NSUInteger, NSMutableIndexSet *> removedItemsBySection;
+  void (^addRemovedIndexPath)(NSIndexPath *) = ^(NSIndexPath *indexPath){
+    const auto &element = removedItemsBySection.find(indexPath.section);
+    if (element == removedItemsBySection.end()) {
+      removedItemsBySection.insert({indexPath.section, [NSMutableIndexSet indexSetWithIndex:indexPath.item]});
+    } else {
+      [element->second addIndex:indexPath.item];
+    }
+  };
+  // Move items
+  [[changeset movedItems] enumerateKeysAndObjectsUsingBlock:^(NSIndexPath *fromIndexPath, NSIndexPath *toIndexPath, BOOL *stop) {
+    insertedItemsBySection[toIndexPath.section][toIndexPath.row] = modifiedSections[fromIndexPath.section][fromIndexPath.item];
+  }];
+  [[changeset movedItems] enumerateKeysAndObjectsUsingBlock:^(NSIndexPath *fromIndexPath, NSIndexPath *toIndexPath, BOOL *stop) {
+    addRemovedIndexPath(fromIndexPath);
+  }];
+  // Remove items
+  for (NSIndexPath *removedItem in [changeset removedItems]) {
+    addRemovedIndexPath(removedItem);
+  }
+  for (const auto &removedItems : removedItemsBySection) {
+    [[modifiedSections objectAtIndex:removedItems.first] removeObjectsAtIndexes:removedItems.second];
+  }
+  // Remove sections
+  [modifiedSections removeObjectsAtIndexes:[changeset removedSections]];
+  // Insert sections
+  [modifiedSections insertObjects:emptyMutableArrays([[changeset insertedSections] count])
+                        atIndexes:[changeset insertedSections]];
+  // Insert items
+  [[changeset insertedItems] enumerateKeysAndObjectsUsingBlock:^(NSIndexPath *indexPath, id model, BOOL *stop) {
+    insertedItemsBySection[indexPath.section][indexPath.item] = model;
+  }];
+  for (const auto &section : insertedItemsBySection) {
+    NSMutableIndexSet *indexes = [NSMutableIndexSet indexSet];
+    NSMutableArray *items = [NSMutableArray array];
+    for (const auto &item : section.second) {
+      [indexes addIndex:item.first];
+      [items addObject:item.second];
+    }
+    [[modifiedSections objectAtIndex:section.first] insertObjects:items atIndexes:indexes];
+  }
+  return [[CKTransactionalComponentDataSourceState alloc] initWithConfiguration:state.configuration
+                                                                       sections:modifiedSections];
+}
+
+static NSArray *emptyMutableArrays(NSUInteger count)
+{
+  NSMutableArray *arrays = [NSMutableArray array];
+  for (NSUInteger i = 0; i < count; i++) {
+    [arrays addObject:[NSMutableArray array]];
+  }
+  return arrays;
+}
+
```

Comment:
Copy/Pasta :(

Likely okay since this is just debugging code but it still feels gross. I'll see if I can simplify this a bit more once I get some more test cases written.